x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java

```diff
@@ -0,0 +1,143 @@
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.master.TransportMasterNodeAction;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.block.ClusterBlockException;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.IndexNotFoundException;
+import org.elasticsearch.repositories.IndexId;
+import org.elasticsearch.repositories.RepositoriesService;
+import org.elasticsearch.repositories.Repository;
+import org.elasticsearch.repositories.RepositoryData;
+import org.elasticsearch.snapshots.SnapshotId;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.elasticsearch.index.IndexModule.INDEX_STORE_TYPE_SETTING;
+
+public class TransportMountSearchableSnapshotAction
+    extends TransportMasterNodeAction<MountSearchableSnapshotRequest, RestoreSnapshotResponse> {
+    // This action doesn't technically need to run on the master node, but it needs to get metadata from the repository and we only expect
+    // the repository to be accessible from data and master-eligible nodes so we can't run it everywhere. Given that we already have a way
+    // to run actions on the master and that we have to do the restore via the master, it's simplest to use TransportMasterNodeAction.
+
+    private final Client client;
+    private final RepositoriesService repositoriesService;
+
+    @Inject
+    public TransportMountSearchableSnapshotAction(TransportService transportService, ClusterService clusterService, Client client,
+                                                  ThreadPool threadPool, RepositoriesService repositoriesService,
+                                                  ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(MountSearchableSnapshotAction.NAME, transportService, clusterService, threadPool, actionFilters,
+            MountSearchableSnapshotRequest::new, indexNameExpressionResolver);
+        this.client = client;
+        this.repositoriesService = repositoriesService;
+    }
+
+    @Override
+    protected String executor() {
+        // Avoid SNAPSHOT since snapshot threads may all be busy with long-running tasks which would block this action from responding with
+        // an error. Avoid SAME since getting the repository metadata may block on IO.
```

Comment:
I think this deserves a tests

Reply:
I'm not sure how to write a test of this that isn't vacuous. Any suggestions?

Reply:
I looked at it and I don't see how to easily craft a good test for this. Maybe just ignore my review comment, as you already commented correctly the usage of this thread pool.

Reply:
(and sorry for the noise)